frontend/src/metabase/query_builder/components/QueryDownloadPopover/use-download-data.ts

```diff
@@ -0,0 +1,80 @@
+import { useAsyncFn } from "react-use";
+import type { AsyncFnReturn } from "react-use/lib/useAsyncFn";
+
+import { useDispatch } from "metabase/lib/redux";
+import { downloadQueryResults } from "metabase/query_builder/actions";
+import type Question from "metabase-lib/v1/Question";
+import type {
+  DashboardId,
+  DashCardId,
+  Dataset,
+  ParameterValuesMap,
+  VisualizationSettings,
+} from "metabase-types/api";
+
+type UseDownloadDataParams = {
+  question: Question;
+  result: Dataset;
+  dashboardId?: DashboardId;
+  dashcardId?: DashCardId;
+  uuid: string | undefined;
+  token: string | undefined;
+  params?: ParameterValuesMap;
+  visualizationSettings?: VisualizationSettings;
+};
+
+export const useDownloadData = ({
+  question,
+  result,
+  dashboardId,
+  dashcardId,
+  uuid,
+  token,
+  params,
+  visualizationSettings,
+}: UseDownloadDataParams): AsyncFnReturn<
+  ({
+    type,
+    enableFormatting,
+  }: {
```

Comment:
nit: you don't have to destructure `type` and `enableFormatting` in the type declaration, you can also do `AsyncFnReturn<(options: {...}) => Promise<void>>